### DISCUSSION TITLE
Copy / Paste Nodes and Edges

### DISCUSSION THREAD

**Discussion by moklick** [xyflow/xyflow#1020] [New Features]
Add the ability to copy/paste nodes and edges with cmd+c, cmd+v and change these keys via `copyKeyCode` and `pasteKeyCode` for example.

**Comment by mikebarkmin** [comment]
I am currently planing to implement this feature for my app. I am thinking about using the clipboard api and defining a custom Content-Type for my app. This should also allow me to copy and paste nodes across flows. Similar to the drag-and-drop example. For this behavior one also needs custom validation to check, if the pasted node is a valid one. I think that another prop like `onNodeValidation` or something similiar would be needed.

onNodeValidation should receive a function with the following definition:

```ts
(node: Node) => boolean;
```

It returns true, if the to be inserted node is valid and and false if not. Only if the node is valid, it will be inserted into the elements array.

**Comment by CarlosGzz** [comment]
there should be a way to copy and paste single and multiple nodes without having to create an external solution. Im already creating a solution since we need this asap but I think a simple implementation only for elements in the canvas should be available.

**Reply by kuldeepkeshwar** [reply]
I also worked on same feature few days back, you check here a demo of it https://discord.com/channels/771389069270712320/859801398228156446/885230876655509596

**Reply by CarlosGzz** [reply]
thanks but we got it :)

**Reply by tabdon** [reply]
@kuldeepkeshwar do you have reference code for that?

**Reply by kuldeepkeshwar** [reply]
https://discord.com/channels/771389069270712320/859801398228156446/900225231975755816

**Comment by kuldeepkeshwar** [comment]
@moklick, I can throw up a pull request for it, if you haven't already started on it

**Reply by maxmatthews** [reply]
@kuldeepkeshwar Any chances you could PR this?

**Comment by Keanottesen** [comment]
What is the status on this?

**Comment by visualsitemaps** [comment]
Hi is this now possible to do with the latest version? if not can someone please share your working fork?
thanks so much!

**Comment by vmwxiong** [comment]
Bumping to say this would be a very much appreciated feature!

**Comment by tminhduc2811** [comment]
Would love to have this feature!

**Comment by pengfu** [comment]
I think this feature is cool!
Recently I implemented  node duplicate by the contextMenu which pops up by right click the node. The interaction  is similar to duplicating a folder on the Mac,  so `copy and paste` and `duplicate ` would be better if both supported.
`Copy / Paste Nodes and Edges` may be means that copy nodes and their edges together.
btw, I am  curious about if we can only copy one edge, it may be useful in some cases because the edge data config work may be expensive. And I wonder whether this idea is feasible in terms of product interaction and technical implementation.  Can we just paste the Independent edge to the canvas ? And I think this is a bit like the feature request post [https://github.com/wbkd/react-flow/issues/2394](https://github.com/wbkd/react-flow/issues/2394)

**Comment by ghost** [comment]
Couldn't implement cmd+c to copy and cmd+v to paste, instead I have select nodes and edges, click on panel, they get duplicated on top of them, now click on it, drag out.

Here is the link to the sample implementation. Note that this is implemented in my project not in react-flow. I think there is no need to copy paste, select the targets, click outside to duplicate is smooth. just my opinion - 

**Implementation Code example** [Select target node and edges, click to paste](https://github.com/ui-editor/flow/pull/7/files#diff-4d8b51992f0a8e8e3c96675c58f6bdde5f18743b676d94b0f2b333c3bcbb9f03R75-R145 )
**Good quality with music-** https://www.youtube.com/watch?v=juRkDY8lEyE
**Quick preview** - 
![react-flow select, click to duplicate](https://user-images.githubusercontent.com/6542274/236194050-7d0f8306-bf2f-47e4-b8cb-e5e1398294c4.gif)

**Reply by MateoWartelle** [reply]
Thank you

**Comment by ghost** [comment]
@moklick checkout above implementation. let me know what you think

**Comment by moklick** [comment]
Copy/Paste/Cut is now available as a pro example https://reactflow.dev/docs/examples/interaction/copy-and-paste/

**Reply by bcakmakoglu** [reply]
Your frustration is understandable but you might want to consider using a more respectful approach if you want people to take your grievances seriously. The library is completely free to use, you are *not* required to *pay anything*.

To keep an open source project like this going resources like money are necessary. 
That is why the xyflow team offers *examples* that can be paid for which showcase how the library can be *used* to solve some use-cases - that means everything in those examples is possible with the library that you can already use *for free*.
There is no hidden pro-library, no additional costs for secret features whatsoever.
You can implement a copy+paste feature easily yourself with all the tools the library offers.

The pro subscription is to *support* this project since GitHub Stars don't translate into money, which, alas, is still necessary to feed and house oneself.
If you don't feel like paying for it then don't, but it's not fair to come here and hurl abuse at wonderful people that contribute a great deal to the open source community.

**Comment by andreasgangso** [comment]
Example hook

```ts
import { ReactFlowInstance, Node } from "@xyflow/react";
import { useCallback, useEffect } from "react";

export const useCopyPaste = (
  rfInstance: ReactFlowInstance | null,
) => {

  const onCopyCapture = useCallback(
    (event: ClipboardEvent) => {
      event.preventDefault();
      const nodes = JSON.stringify(
        rfInstance?.getNodes().filter((n) => n.selected)
      );

      event.clipboardData?.setData("flowchart:nodes", nodes);
    },
    [rfInstance]
  );

  const onPasteCapture = useCallback(
    (event: ClipboardEvent) => {
      event.preventDefault();
      const nodes = JSON.parse(
        event.clipboardData?.getData("flowchart:nodes") || "[]"
      ) as Node[] | undefined;
      if (nodes) {
        const randomId = () => Math.random().toString(16).slice(2);
        rfInstance?.setNodes(
          [
            ...rfInstance.getNodes().map((n) => ({ ...n, selected: false })),
            ...nodes.map((n) => ({
              ...n,
              selected: true,
              id: randomId(),
              position: { x: n.position.x + 10, y: n.position.y + 10 },
          }))
        ]);
      }
    },
    [rfInstance]
  );

  useEffect(() => {
    window.addEventListener("copy", onCopyCapture);
    return () => {
      window.removeEventListener("copy", onCopyCapture);
    };
  }, [onCopyCapture]);

  useEffect(() => {
    window.addEventListener("paste", onPasteCapture);
    return () => {
      window.removeEventListener("paste", onPasteCapture);
    };
  }, [onPasteCapture]);
};

```
